Explorer 2.0

abrooksv opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
* Uses AwsResourceCache
* Uses AsyncTreeModel to remove custom loading node logic
* Refreshing the tree does not require replacing the Tree Swing object
* Adds in a CopyArn action
* Removes pagination

## Motivation and Context
Cleans up and simplifies the explore rlogic

## Related Issue(s)
<!--- What is the related issue you are trying to fix? -->

## Testing
Mostly manual, needs unit tests written

## Screenshots (if appropriate)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt:
```diff
@@ -29,27 +30,32 @@ import software.aws.toolkits.jetbrains.core.credentials.ProjectAccountSettingsMa
-    private var awsTree: Tree? = null
+    private val awsTreeModel = AwsExplorerTreeStructure(project)
+    private val structureTreeModel = StructureTreeModel(awsTreeModel, project)
+    private var awsTree = createTree(AsyncTreeModel(structureTreeModel, true, project))
```
Thought we didn't have to replace the tree, why does this have to be a `var`?

abrooksv replied:
It can be a val now

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt:
```diff
@@ -65,70 +71,67 @@ class ExplorerToolWindow(private val project: Project) : SimpleToolWindowPanel(t
         project.messageBus.connect().subscribe(ProjectAccountSettingsManager.ACCOUNT_SETTINGS_CHANGED, this)
 
-        updateModel()
+        treePanelWrapper.setContent(errorPanel)
```
Is `errorPanel` the "select creds please" panel or something else?

abrooksv replied:
You are correct

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt:
```diff
@@ -65,70 +71,67 @@ class ExplorerToolWindow(private val project: Project) : SimpleToolWindowPanel(t
+    }
 
-        awsTree = newTree
+    internal fun refreshTree() {
```
Consider renaming to `invalidateTree` to match the underlying method?

abrooksv replied:
Done

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/ExplorerToolWindow.kt:
```diff
@@ -65,70 +71,67 @@ class ExplorerToolWindow(private val project: Project) : SimpleToolWindowPanel(t
-                    awsExplorerNode.onDoubleClick(model, selectedElement)
-                }
+
+        object : DoubleClickListener() {
```
So weird we have to wire this manually

abrooksv replied:
Ya but new way is cleaner. :D

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/actions/CopyArn.kt:
```diff
@@ -0,0 +1,15 @@
+import com.intellij.openapi.ide.CopyPasteManager
+import java.awt.datatransfer.StringSelection
+
+class CopyArn(private val arn: String) : AnAction("Copy ARN") {
```
Prly want a `message` for this string.

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/nodes/AwsExplorerNode.kt:
```diff
@@ -40,5 +38,9 @@ abstract class AwsExplorerNode<T>(val nodeProject: Project, value: T, private va
+    open fun onDoubleClick() {}
+
+    override fun toString(): String {
+        return value.toString()
```
Surprised ktlint didn't kick off about this - it could be a block expression.

abrooksv replied:
It probably will. More commits with come to fix the tests and do a few small things. There are regressions in behavior

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/resources/LambdaResources.kt:
```diff
@@ -0,0 +1,50 @@
+        CloudFormationCachedResource { listStacksPaginator().stackSummaries().toList() }
+    }
+
+    fun listStacks() = listStacks
```
Any reason not to just expose this as a `LIST_STACKS` constant val?

abrooksv replied:
Wanted some consistentcy, that's all

kiiadi replied:
Consistency with what? https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/resources/LambdaResources.kt#L12

abrooksv replied:
Just within this class 😄 
Mixing consts and functions was all. I dont have strong opinion

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/resources/LambdaResources.kt:
```diff
@@ -0,0 +1,50 @@
+        CloudFormationStackResourcesCachedResource(stackId)
+
+    private class CloudFormationCachedResource<T>(private val call: CloudFormationClient.() -> T) :
+        CachedResourceBase<T, CloudFormationClient>(CloudFormationClient::class) {
```
In the view iteration this becomes a bit less clunky.

abrooksv replied:
Sounds good to me.

kiiadi commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/resources/LambdaResources.kt:
```diff
@@ -0,0 +1,50 @@
+
+    fun listStacks() = listStacks
+
+    fun listStackResources(stackId: String): CachedResource<List<StackResourceSummary>> =
```
I think I need to add an explicit id to each cached resources because I'm concerned that the backing cache is using object identity - which obviously means that the cache will never get hit. Will investigate and update.

abrooksv replied:
Ah I guessed it was equals based due to you used a data class as the key...I think....

abrooksv commented on jetbrains-core/resources/META-INF/plugin.xml:
```diff
@@ -161,8 +161,7 @@
                         testServiceImplementation="software.aws.toolkits.jetbrains.core.MockClientManager"/>
         <projectService serviceInterface="software.aws.toolkits.jetbrains.core.AwsResourceCache"
-                        serviceImplementation="software.aws.toolkits.jetbrains.core.DefaultAwsResourceCache"
-                        testServiceImplementation="software.aws.toolkits.jetbrains.core.MockResourceCache"/>
```
Not if I can help it 😈

abrooksv commented on core/src/software/aws/toolkits/core/utils/LogUtils.kt:
```diff
@@ -6,8 +6,10 @@ package software.aws.toolkits.core.utils
+import kotlin.reflect.KClass
 
 inline fun <reified T : Any> getLogger(): Logger = LoggerFactory.getLogger(T::class.java)
+fun getLogger(clazz: KClass<*>): Logger = LoggerFactory.getLogger(clazz.java)
```
Ya, good idea

abrooksv commented:
I was re-basing to re-add the mock resource cache...but after the conversation today looking at removing it...again :D

The implementation itself is done, it is just tests. I cleaned up CFN Nodes to also use cache for resource level too. Also switched to `getResourceNow`
